core/src/main/scala/org/apache/spark/TaskEndReason.scala

```diff
@@ -115,22 +118,34 @@ case class ExceptionFailure(
     description: String,
     stackTrace: Array[StackTraceElement],
     fullStackTrace: String,
-    metrics: Option[TaskMetrics],
-    private val exceptionWrapper: Option[ThrowableSerializationWrapper])
+    exceptionWrapper: Option[ThrowableSerializationWrapper],
+    accumUpdates: Seq[AccumulableInfo] = Seq.empty[AccumulableInfo])
   extends TaskFailedReason {
 
+  // For backward compatibility; this was replaced with `accumUpdates` in Spark 2.0.
+  val metrics: Option[TaskMetrics] = {
+    if (accumUpdates.nonEmpty) {
```

Comment:
I appreciate that you didn't try to cram this all into a dense confusing one-liner :+1: